### PR TITLE
Updates based on genius.it discussion

### PR DESCRIPTION
**Pull request by devd**
Quite a mishmash of stuff. Noted down issues in places where I didn't end up doing what we discussed.

### CODE REVIEW

**Comment by devd** [comment]
@metromoxie take a look!

File: index.bs [hunk 1]
```diff
@@ -107,6 +107,15 @@ because they are by definition unique origins, with no relationship to the
 original origin, designing permissions for them to access resources of the
 original origin would be difficult.
 
+Issue: TODO: Problems with CSP sandbox goes well beyond this. synthetic origins
```
Comment (metromoxie) [inline index.bs]: All great points. Maybe file an issue to track this as well?

File: index.bs [hunk 2]
```diff
@@ -127,57 +136,47 @@ companies with large portfolios of disparate products all under the same domain.
...
-Issue: Not sure how to actually refer to 'real origins'. This is a terrible
-name, and we need a better way to talk about them. Maybe physical origin?
-Traditional origin? (jww)
-
 ## Use Cases/Examples ## {#usecases}
 
 We see effectively three different use cases for Per-page Suborigins:
 
-1. Separating distinct applications that happen to be served from the same
+1. Separating distinct applications served, due to deployment issues, from the same
```
Comment (metromoxie) [inline index.bs]: A little awkward. Maybe rephrase as "Separating distinct applications that are served from the same domain because of deployment issues, but do not need..."?

File: index.bs [hunk 3]
```diff
@@ -340,6 +341,10 @@ document on
 the attacker should not be able to control any document on
 `example.com` not in the `foo` namespace.
 
+Issue: TODO(devd): Should we also assert that attacker on main/parent origin
```
Comment (metromoxie) [inline index.bs]: Since I see the relationship as symmetric, I think we should. Feel free to add that in now, or leave this issue in for later.

File: index.bs [hunk 4]
```diff
@@ -364,58 +369,62 @@ to ease the use of Suborigins in legacy applications are also presented. See
 ## Representation of Suborigins ## {#representation}
 
 At an abstract level, a suborigin consists of the <a>scheme</a>, <a>host</a>,
-and <a>port</a> of a traditional origin, plus a <a>suborigin namespace</a>.
+and <a>port</a> of a physical (traditional) origin, plus a <a>suborigin namespace</a>.
```
Comment (metromoxie) [inline index.bs]: I think you should remove the "(traditional)" parentheticals. I added #17 so that we eventually define 'physical origin' in the spec, and then we can make these all references to those.

File: index.bs [hunk 5]
```diff
@@ -364,58 +369,62 @@ to ease the use of Suborigins in legacy applications are also presented. See
...
 
-Similarly, if a resource is hosted at `https://example.com:8080/` in
-the suborigin namespace `separate`, this would be serialized as
+Similarly, a resource hosted at `https://example.com:8080/` in
+the suborigin namespace `separate` would be serialized as
 `https://separate_example.com:8080/`.
 
-Internally, the <a>suborigin namespace</a> must be tracked by the user agent.
-When the origin needs to be serialized for a resource, the user agent should
+Internally, the user agent just track the <a>suborigin namespace</a> of the resource.
```
Comment (metromoxie) [inline index.bs]: track -> tracks

File: index.bs [hunk 6]
```diff
@@ -364,58 +369,62 @@ to ease the use of Suborigins in legacy applications are also presented. See
...
+When the origin needs to be serialized, the user agent should
 follow the algorithm in [[#serializing]].
 
+Note that the underscore character is not a valid character for hostnames used
+on the web [[!RFC1123]] and as a result, the serialization above cannot collide
```
Comment (metromoxie) [inline index.bs]: "cannot" is a bit strong; the Web is a scary and big place :-) Maybe we should rephrase as "should not" or, alternatively, "cannot collide with valid existing hostnames"?

File: index.bs [hunk 7]
```diff
@@ -364,58 +369,62 @@ to ease the use of Suborigins in legacy applications are also presented. See
...
-server (rather than a resource itself) is treated as the only authoritative
-source of the suborigin namespace of a resource. This is implemented through an
-additional header-only Content Security Policy directive `suborigin`, which
-takes a string value that is the namespace. For example, to put a resource in
-the `testing` suborigin namespace, the server would specify the following
-directive in the CSP header:
+isolation suborigins intend to provide. To prevent this, the
+server (rather than a resource itself) is the only authoritative
+source of the suborigin namespace of a resource. The server communicates the
+suborigin of a resource to the user agent through a new `suborigin` header,
```
Comment (metromoxie) [inline index.bs]: Can you add a link from "suborigin header" to the definition of that header?

File: index.bs [hunk 8]
```diff
@@ -364,58 +369,62 @@ to ease the use of Suborigins in legacy applications are also presented. See
...
 
-## The `suborigin` Directive ## {#the-suborigin-directive}
+## The `suborigin` header ## {#the-suborigin-header}
 
-Suborigins are defined by a <dfn>suborigin directive</dfn> in the Content
-Security Policy [[!CSP2]] of the resource. The syntax for the name and value of
-the directive are described by the following ABNF grammar:
+Suborigins are defined by a <dfn>suborigin</dfn> HTTP response header. The syntax
+for the name and value of the header are described by the following ABNF
+grammar (RFC5234):
```
Comment (metromoxie) [inline index.bs]: Please change "(RFC5234)" to a normative link to the RFC (I think "[[!RFC5234]]" should do the trick).

File: index.bs [hunk 9]
```diff
@@ -437,6 +446,9 @@ do support suborigins. The following discussions discuss the three major
 cross-origin mechanisms that are relevant: <a>CORS</a>, <a>`postMessage`</a>,
 and Workers [[!WORKERS]].
 
+Issue: TODO(devd): Making things specific to XHR or CORS is weird. We should
```
Comment (metromoxie) [inline index.bs]: Isn't that, more or less, what I wrote? Is it just perhaps the heading that you don't like? Should we rename it "Fetching" instead of "CORS"?

File: index.bs [hunk 10]
```diff
@@ -456,37 +468,30 @@ origin, as described in [[#serializing]].
 
 Issue: TODO(jww): Formal definition of the headers and responses w/grammars.
 Also need to be explicit about `*` having same limitations as
-`Access-Control-Allow-Origin` w/credentials. Also need to describe opt-outs to
-use old-school Origin header and ACAO header.
+`Access-Control-Allow-Origin` w/credentials.
```
Comment (metromoxie) [inline index.bs]: Why did you remove the rest of this TODO?

File: index.bs [hunk 11]
```diff
@@ -456,37 +468,30 @@ origin, as described in [[#serializing]].
...
-
-The Service Worker [[!SERVICE-WORKERS]] registration API must not be accessible
-from suborigins. As a result of the above restriction on Workers, since a
-Service Worker cannot be created by a Suborigin, no Service Workers will be able
-to intercept the requests of a Suborigin.
 
 Note: This may change in the future, and Suborigins may eventually be allowed to
-register Service Workers, but for now, allowing the creation of Service Workers
-from Suborigins adds too many complications.
+register Service Workers, but for now, allowing the creation of (Service)
```
Comment (metromoxie) [inline index.bs]: I find "(Service) Workers" awkward. Maybe "all Workers, including Service Workers and Shared Workers," instead?

File: index.bs [hunk 12]
```diff
@@ -531,11 +533,9 @@ The suborigin of a resource is the value computed by the following algorithm:
   </li>
 
   <li>
-    If the Content Security Policy of the resource contains a valid
-    <a>suborigin directive</a> in the <a
-    href="https://www.w3.org/TR/CSP2/#policy-syntax">directive list</a>
-    [[!CSP2]], then let `suborigin-namespace` be the
-    `directive-value`.
+    If the HTTP response of the resource contains a valid
+    <pre>suborigin</pre> header, then let `suborigin-namespace` be the
```
Comment (metromoxie) [inline index.bs]: <pre>suborigin</pre> -> [suborigin]

or if you want to keep the code look, [`suborigin`] should also work.

Bikeshed FTW!

File: index.bs (hunk 10 above)
Comment (metromoxie) [inline index.bs]: n/m, I see you moved it below :-)

**Comment by metromoxie** [comment]
These are fantastic changes! Mostly minor comments. Also, if you can compile before merging, too, that would be great.

File: index.bs (hunk 9 above)
Comment (devd) [inline index.bs]: no, because once you make all fetches crossorigins, then even SRI script tag fetches are cross origin?

File: index.bs (hunk 11 above)
Comment (devd) [inline index.bs]: hmm .. it looked weird to phrase it that way in a "Note:". I thought (Service) works better for a non-normative note. wdyt?

**Comment by devd** [comment]
I was thinking of just pushing this and then go through and file issues for anything that is needed in a separate step.

File: index.bs (hunk 9 above)
Comment (metromoxie) [inline index.bs]: I'm still confused, so let's add the TODO for now and discuss it later :-)

File: index.bs (hunk 11 above)
Comment (metromoxie) [inline index.bs]: Normative or not, the "(Service)" reads awkwardly to me, and also seems to exclude Shared Workers.

**Comment by metromoxie** [comment]
That sgtm as a plan.

File: index.bs (hunk 10 above)
Comment (devd) [inline index.bs]: I was thinking that should go into the whole opt-outs section, rather than here (definition of headers etc)?

Comment (metromoxie) [inline index.bs]: lgtm
